### PR TITLE
Graph save/restore actions

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
This PR introduces new graph actions:
 - [x] save_to() - saves all trainable modules in a given graph to checkpoint
 - [x] restore_from() - restores all trainable modules from a checkpoint

Additionally, as Neural Graphs are supposed to be "backend-dependent", it polishes a bit the way we inject the "backend dependencies", by:
 - [x] introducing the "import_backend" solution based on the Keras approach
 - [x] introducing backend proxies for: save, load, get_state_dict, set_state_dict

### CODE REVIEW

File: nemo/backends/torch_backend.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+from os.path import expanduser
+from typing import Any, Dict
+
+import torch
+
+
+def save(checkpoint: Dict[str, Any], filename: str) -> None:
+    """
+        A proxy function that saves the checkpoint to a given file.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:25]: Indentation should be reduced by 1 tab size.

File: nemo/backends/torch_backend.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            checkpoint - checkpoint to be stored.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:27]: Capitalize `checkpoint` to be consistent with `Name...` below

File: nemo/backends/torch_backend.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            filename - Name of the file containing checkpoint.
+    """
+    # Greate the absolute path and save.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:30]: "Greate" -> "Get"

File: nemo/backends/torch_backend.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    abs_filename = expanduser(filename)
+    torch.save(checkpoint, abs_filename)
+
+
+def load(filename: str) -> Dict[str, Any]:
+    """
+         A proxy function that loads checkpoint from a given file.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:37]: Indentation

File: nemo/backends/torch_backend.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            filename - Name of the file containing checkpoint.
+        Returns:
+            Loaded checkpoint.
+    """
+    # Greate the absolute path and save.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:43]: "Get"

File: nemo/backends/torch_backend.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    abs_filename = expanduser(filename)
+    # Use map location to be able to load CUDA-trained modules on CPU.
+    return torch.load(abs_filename, map_location=lambda storage, loc: storage)
+
+
+def get_state_dict(model: torch.nn.Module) -> Dict[str, Any]:
+    """
+        A proxy function that gets the state dictionary.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:51]: Indentation

File: nemo/backends/torch_backend.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            model - torch model.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:53]: Capitalize "torch" to be consistent with above.

File: nemo/backends/torch_backend.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    return model.state_dict()
+
+
+def set_state_dict(model: torch.nn.Module, state_dict: Dict[str, Any]) -> None:
+    """
+        A proxy function that sets the state dictionary.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:60]: Indentation

File: nemo/backends/torch_backend.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            model - torch model.
```
Comment (titu1994) [inline nemo/backends/torch_backend.py:62]: Capitalize "torch" to be consistent as above

File: nemo/core/neural_graph.py [hunk 10]
```diff
@@ -962,3 +967,79 @@ def unfreeze(self, module_names: Optional[List[str]] = None):
             if module.type == ModuleType.trainable:
                 # Unfreeze weights of the module.
                 module.unfreeze()
+            else:
+                logging.debug("Module `{}` is not trainable so cannot be unfrozen".format(name))
+
+    def save_to(self, filename: str, module_names: Optional[List[str]] = None):
+        """
+            Saves the state of trainable modules in the graph to a checkpoint file.
```
Comment (titu1994) [inline nemo/core/neural_graph.py:975]: Indentation

File: nemo/core/neural_graph.py [hunk 11]
```diff
@@ -962,3 +967,79 @@ def unfreeze(self, module_names: Optional[List[str]] = None):
...
+            Raises:
+                KeyError: If name of the module won't be recognized.
+        """
+        # Work on all modules.
+        if module_names is None:
+            module_names = self._modules.keys()
+
+        # Prepare the "graph checkpoint".
+        chkpt = {"header": {}, "modules": {}}
+        chkpt["header"] = {"nemo_core_version": nemo_version, "name": self.name}
```
Comment (titu1994) [inline nemo/core/neural_graph.py:988]: This can be inlined above to avoid empty dictionary allocation.

File: nemo/core/neural_graph.py [hunk 12]
```diff
@@ -962,3 +967,79 @@ def unfreeze(self, module_names: Optional[List[str]] = None):
...
+                logging.debug("Module `{}` is not trainable so cannot be saved".format(name))
+
+        # Save checkpoint.
+        save(chkpt, filename)
+        log_str = "Saved  the '{}' graph to a checkpoint `{}`:\n".format(self.name, filename) + log_str
+        logging.info(log_str)
+
+    def restore_from(self, filename: str, module_names: Optional[List[str]] = None):
+        """
+            Restores the state of trainable modules in the graph from a checkpoint file.
```
Comment (titu1994) [inline nemo/core/neural_graph.py:1011]: Indentation

File: nemo/core/neural_graph.py [hunk 13]
```diff
@@ -962,3 +967,79 @@ def unfreeze(self, module_names: Optional[List[str]] = None):
...
+        for name in module_names:
+            try:
+                # Get module.
+                module = self._modules[name]
+                # Restore module weights
+                set_state_dict(module, chkpt["modules"][name])
+                log_str += "  * Module '{}' ({}) params loaded\n".format(module.name, type(module).__name__)
+            except KeyError:
+                log_str += "  ! Module '{}' params not found in checkpoint\n".format(name)
+                warning = True
```
Comment (titu1994) [inline nemo/core/neural_graph.py:1039]: Default value of `warning` needs to be declared outside the loop, in the edge case that warning is never defined and the loop completes and the below `if warning:` check executes.

File: tests/unit/core/neural_graph/test_neural_graphs.py [hunk 14]
```diff
@@ -83,18 +85,35 @@ def test_explicit_graph_manual_activation(self):
         assert len(g0) == 2
 
     @pytest.mark.unit
-    def test_default_output_ports(self):
-        """ Tests automatic binding of default output ports. """
+    def test_graph_save_load(self, tmpdir):
+        """
+            Tests graph saving and loading.
```
Comment (titu1994) [inline tests/unit/core/neural_graph/test_neural_graphs.py:90]: Indentation

File: tests/unit/test_torch_backend.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+from nemo.backends import get_state_dict, load, save, set_state_dict
+from nemo.backends.pytorch.tutorials import TaylorNet
+
+
+@pytest.mark.usefixtures("neural_factory")
+class TestTorchBackend:
+    @pytest.mark.unit
+    def test_state_dict(self):
+        """
+            Tests whether the get/set_state_dict proxy functions work properly.
```
Comment (titu1994) [inline tests/unit/test_torch_backend.py:32]: Indentation

File: tests/unit/test_torch_backend.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+        # Compare state dicts.
+        state_dict2 = get_state_dict(fx)
+        for key in state_dict1.keys():
+            assert array_equal(state_dict1[key].cpu().numpy(), state_dict2[key].cpu().numpy())
+
+    @pytest.mark.unit
+    def test_save_load(self, tmpdir):
+        """
+            Tests whether the save and load proxy functions work properly.
```
Comment (titu1994) [inline tests/unit/test_torch_backend.py:51]: Indentation

File: nemo/backends/torch_backend.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/backends/torch_backend.py:25]: ech, but I like this indent in comments... ;)

File: nemo/backends/torch_backend.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/backends/torch_backend.py:27]: ok

File: nemo/backends/torch_backend.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/backends/torch_backend.py:30]: goot catch!

File: nemo/core/neural_graph.py (hunk 11 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_graph.py:988]: ok. But not sure how important that is...

File: nemo/core/neural_graph.py (hunk 13 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_graph.py:1039]: of course! good catch! :)

File: nemo/backends/torch_backend.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/backends/torch_backend.py:25]: done
